ambry-store/src/main/java/com/github/ambry/store/ReplicaPlacementValidator.java

```diff
@@ -0,0 +1,146 @@
+          brokenDisks.add(currentDisk);
+        }
+      }
+    }
+  }
+
+  /**
+   * Check the placement of replicas on the disks and reshuffle them if necessary.
+   * @return A map of the broken disks and the disks they should be reshuffled to.
+   *         An empty map if no reshuffling is necessary or if the reshuffling failed.
+   */
+  public Map<DiskId, DiskId> reshuffleDisks() {
+    Map<DiskId, DiskId> shuffledDisks = new HashMap<>();
+
+    // Sanity checks: - Abort if we did not find the same number of disks on the
+    //                  host as we got from Helix.
+    //                - Abort if we did not find any broken disks.
+    if ((expectedDiskToReplicaMap.size() != foundDiskToPartitionMap.size()) ||
+        brokenDisks.size() == 0) {
+      return Collections.emptyMap();
+    }
+
+    for (DiskId currentDisk : brokenDisks) {
+      List<ReplicaId> expectedReplicas = expectedDiskToReplicaMap.get(currentDisk);
+      DiskId foundDisk = findDiskWithReplicas(expectedReplicas);
+      if(foundDisk == null) {
+        return Collections.emptyMap();
+      } else {
+        shuffledDisks.put(currentDisk, foundDisk);
+      }
+    }
+    return shuffledDisks;
+  }
+
+  /**
+   * Find the disk that contains the expected replicas.
+   * @param expectedReplicas A list of replicas that should be on the disk.
+   * @return The disk that contains the expected replicas, or null if no such disk was found.
+   */
+  private DiskId findDiskWithReplicas(List<ReplicaId> expectedReplicas) {
+    for (Map.Entry<DiskId, Set<String>> entry : foundDiskToPartitionMap.entrySet()) {
+      DiskId currentDisk = entry.getKey();
+      Set<String> partitions = entry.getValue();
+      boolean found = true;
+      for (ReplicaId replica : expectedReplicas) {
+        String partitionID = replica.getPartitionId().toString();
+        if(!partitions.contains(partitionID)) {
+          found = false;
+          break;
+        }
+      }
+      if(found) {
+        return currentDisk;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Checks whether a directory name looks like a partition.
+   * @param directoryName the name of the directory to check.
+   * @return True if the directory name looks like a partition, false otherwise.
+   */
+  private boolean looksLikePartition(String directoryName) {
+    try {
+      // The convention is simply to use Java long as partition IDs.
+      Long.parseLong(directoryName);
+    } catch (NumberFormatException e) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Find all the partition directories on the disk.
+   * @param disk an instance of DiskId to search for partition directories.
+   * @return A list of partition directories on the disk.
+   */
+  private Set<String> findPartitionsOnDisk(DiskId disk) {
+    Set<String> partitionDirs = new HashSet<>();
+    File[] directories = new File(disk.getMountPath()).listFiles(File::isDirectory);
```

Comment:
We should probably check if the `disk.getMountPath()` exists or not. List file on nonexisting directory would throw some exception.

Reply:
Yeah, thanks - I added error handling for this.